enhancements/ingress/haproxy-healthcheckinterval.md

```diff
@@ -0,0 +1,301 @@
+---
+title: global-healthcheck-interval-in-haproxy
+authors:
+  - "@cholman"
+reviewers:
+  - "@frobware"
+  - "@knobunc"  
+  - "@Miciah"
+  - "@miheer"
+  - "@rfredette"
+  - "@deads2k"
+approvers:
+  - "@frobware"
+  - "@Miciah"
+api-approvers: # in case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers)
+  - "@deads2k"
+creation-date: 2021-11-05
+last-updated: 2021-11-05
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Global Healthcheck Interval in HAProxy
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Expose and make configurable the ROUTER_BACKEND_CHECK_INTERVAL environment variable in HAProxy's template so that
+administrators may customize the length of time between subsequent healthchecks of backend services.
+
+This is already configurable via a route annotation called `router.openshift.io/haproxy.health.check.interval`, but
+exposing the healthcheck interval at a global scope is desired for efficient administration of routes.  HAProxy allows
+setting the healthcheck globally as well as per-route, and both options will be addressed as a part of this proposal.
+
+## Motivation
+
+Too frequent healthcheck probes have been identified as a cause of concern for customers with highly loaded clusters. The
+default healthcheck interval is set to 10 seconds.  For high-load customers, the ability to reduce the frequency of the
+backend healthcheck interval, without having to configure each route separately, is an important factor.
+
+### Goals
+
+Enable the configuration of a backend healthcheck interval via the `IngressControllerSpec`, specifically via the
+`IngressControllerSpecTuningOptions`, with a new parameter `HealthCheckInterval`, which represents a duration of time.
+`HealthCheckInterval` is exposed as the environment variable `ROUTER_BACKEND_CHECK_INTERVAL`.
+
+### Non-Goals
+
+Although the frequency of the healthcheck interval is a cause of concern for some customers, this has not been
+identified as a cause for concern for all customers, chiefly because the default is 10 seconds, which is documented in
```

Comment:
The complete timeout is the sum of the read timeout + connect timeout.  See:
https://github.com/openshift/enhancements/pull/952/files#diff-0b69ed9fb9fcb39200a3b4bbcb5da3f654f92af1530453835f5eaabdd5083097R97-R101

If `timeout check` is set, HAProxy "uses min(`timeout connect`, `inter`) as a connect timeout for check and `timeout check` as an additional read timeout"
- if `timeout check` is not set, HAProxy "uses `inter` for the complete check timeout (connect + read)"

We use the `timeout check 5000ms` in the haproxy config for plain http backend, backend with TLS terminated at the edge, secure backend with re-encryption, and for passthrough.

So `timeout connect 5s` is global, but the total timeout is the sum of `timeout connect` and `timeout check` when the latter is set.

Reply:
> The complete timeout is the sum of the read timeout + connect timeout.

That makes sense.  However, the text in the enhancement seems to be saying that the default healthcheck interval is 10 seconds:

> Although the frequency of the healthcheck interval is a cause of concern for some customers, this has not been
identified as a cause for concern for all customers, chiefly because the default is 10 seconds

Reply:
If we used all the defaults, we have `timeout connect 5s` + `timeout check 5s` = 10s for plain http backend, backend with TLS terminated at the edge, secure backend with re-encryption, and for passthrough.  Do you mean I have to qualify that it is only for those backend types?

Reply:
No, I mean that the total _timeout_ might be 10 seconds, but the _interval_ is 5 seconds whereas the enhancement seems to be saying that the interval is 10 seconds.

Reply:
`timeout check` is always set, so we don't use `check inter`'s value

Reply:
I see where I misunderstood here, conflating timeouts with interval.  Thanks for persisting with me.  Will post an update